docs-src/0.4/en/contributing/index.md

```diff
@@ -19,33 +19,59 @@ If you've fixed [an open issue](https://github.com/DioxusLabs/dioxus/issues), fe
 All pull requests (including those made by a team member) must be approved by at least one other team member.
 Larger, more nuanced decisions about design, architecture, breaking changes, trade-offs, etc. are made by team consensus.
 
-## Tools
+## Before you contribute
 
-The following tools can be helpful when developing Dioxus. Many of these tools are used in the CI pipeline. Running them locally before submitting a PR instead of waiting for CI can save time.
+You might be surprised that a lot of checks fail when making your first PR.
+That's why you should first run these commands before contributing, and it will save you *lots* of time, because the
+GitHub CI is much slower at executing all of these than your PC.
 
-- All code is tested with [cargo test](https://doc.rust-lang.org/cargo/commands/cargo-test.html)
+- Format code with [rustfmt](https://github.com/rust-lang/rustfmt):
 
 ```sh
 cargo fmt --all
 ```
 
-- All code is formatted with [rustfmt](https://github.com/rust-lang/rustfmt)
+- Check all code [cargo check](https://doc.rust-lang.org/cargo/commands/cargo-check.html):
 
 ```sh
 cargo check --workspace --examples --tests
 ```
 
-- All code is linted with [Clippy](https://doc.rust-lang.org/clippy/)
+- Check if [Clippy](https://doc.rust-lang.org/clippy/) generates any warnings. Please fix these!
 
 ```sh
 cargo clippy --workspace --examples --tests -- -D warnings
 ```
 
-- Crates that use unsafe are checked for undefined behavior with [MIRI](https://github.com/rust-lang/miri). MIRI can be helpful to debug what unsafe code is causing issues. Only code that does not interact with system calls can be checked with MIRI. Currently, this is used for the two MIRI tests in `dioxus-core` and `dioxus-native-core`.
+- Test all code with [cargo-test](https://doc.rust-lang.org/cargo/commands/cargo-test.html):
+
+```sh
+cargo test --all --tests
+```
+
+- More tests, this time with [cargo-make](https://sagiegurari.github.io/cargo-make/). Here are all steps, including installation:
+
+```sh
+cargo install --force cargo-make
+cargo make tests
+```
+
+<!-- Playwright doesn't work on my PC despite using the exact same commands as in playwright.yml.
```

Comment:
What error did you run into when running these commands?

Reply:
Oh, on mac if you copy the comment `cd` fails which causes the playwright to fail

Reply:
> This works on my MacOS computer. You do need to have npm installed

I'm on Windows with NPM also installed.

> What error did you run into when running these commands?

This is the full log:
```
[WebServer]     Finished dev [unoptimized + debuginfo] target(s) in 0.60s
[WebServer]      Running `C:\Users\aurel\source\repos\dioxus_forks\dioxus\target\debug\dioxus-playwright-liveview-test.exe`
[WebServer] Listening on http://127.0.0.1:3030
[WebServer]     Finished dev [unoptimized + debuginfo] target(s) in 1.12s
[WebServer]      Running `C:\Users\aurel\source\repos\dioxus_forks\dioxus\target\debug\dx.exe serve`
[WebServer] [WARN] You appear to be creating a Dioxus project from scratch; we will use the default config
[WebServer] [INFO] 🚅 Running build command...
[WebServer] [INFO] 👑 Build done.
[WebServer] [WARN] failed to parse `name` custom section Invalid name type (at offset 47361221)
[WebServer] [INFO] 🚀 Starting development server...
[WebServer]
Dioxus @ v0.4.1 [12:29:50]

        > Local : http://localhost:8080/
        > Network : http://192.168.0.165:8080/
        > HTTPS : Disabled

        > Profile : Debug
        > Hot Reload : Normal
        > Index Template : Default
        > URL Rewrite [index_on_404] : True

[WebServer]     > Build Time Use : 2890 millis

[INFO] A perfect compilation!

[WebServer]     Finished dev [unoptimized + debuginfo] target(s) in 0.86s
[WebServer]      Running `C:\Users\aurel\source\repos\dioxus_forks\dioxus\target\debug\dx.exe build --features web --release`
[WebServer] [WARN] You appear to be creating a Dioxus project from scratch; we will use the default config
[WebServer] [INFO] 🚅 Running build command...
[WebServer] [INFO] 👑 Build done.
[WebServer] [WARN] failed to parse `name` custom section Invalid name type (at offset 3250845)
Error: Process from config.webServer exited early.







To open last HTML report run:

  npx playwright show-report
```
Opening the report shows the report page, but there's no tests that ran.

Reply:
> Oh, on mac if you copy the comment cd fails which causes the playwright to fail

Removed the comment.

Reply:
![noresults](https://github.com/DioxusLabs/docsite/assets/76173380/a7e5691c-3758-4cf5-b736-ffb22741f53c)

I guess that's it then haha. I'll add a disclaimer that it might not work on their PC (for some reason).

Reply:
I think it means one of these commands failed: https://github.com/DioxusLabs/dioxus/blob/master/playwright-tests/playwright.config.js#L74

Do each of these commands work individually on your computer?
```sh
cargo run --package dioxus-playwright-liveview-test --bin dioxus-playwright-liveview-test
cargo run --package dioxus-cli -- serve
cargo run --package dioxus-cli -- build --features web --release\ncargo run --release --features ssr
```

Reply:
I'm not sure why playwright wouldn't show the error/the command wouldn't throw an error if one of them did fail

Reply:
> cargo run --package dioxus-cli -- serve

That fails. Full error log (I've got the `wasm32-unknown-unknown` target installed):

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.99s
     Running `target\debug\dx.exe serve`
[WARN] You appear to be creating a Dioxus project from scratch; we will use the default config
[INFO] 🚅 Running build command...
\ 💼 Waiting to start build the project...
error: no bin target named `dioxus_app`.

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: failed reading 'C:\Users\aurel\source\repos\dioxus_forks\dioxus\target\wasm32-unknown-unknown/debug\dioxus_app.wasm'

Caused by:
    The system cannot find the file specified. (os error 2)', packages\cli\src\builder.rs:129:14
stack backtrace:
   0:     0x7ff6bc9e1e1c - std::sys_common::backtrace::_print::impl$0::fmt
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\sys_common\backtrace.rs:44
   1:     0x7ff6bca0d0eb - core::fmt::rt::Argument::fmt
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\core\src\fmt\rt.rs:138
   2:     0x7ff6bca0d0eb - core::fmt::write
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\core\src\fmt\mod.rs:1094
   3:     0x7ff6bc9da5cf - std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\io\mod.rs:1714
   4:     0x7ff6bc9e1bcb - std::sys_common::backtrace::_print
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\sys_common\backtrace.rs:47
   5:     0x7ff6bc9e1bcb - std::sys_common::backtrace::print
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\sys_common\backtrace.rs:34
   6:     0x7ff6bc9e487a - std::panicking::default_hook::closure$1
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\panicking.rs:269
   7:     0x7ff6bc9e44cf - std::panicking::default_hook
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\panicking.rs:288
   8:     0x7ff6bc9e4f2e - std::panicking::rust_panic_with_hook
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\panicking.rs:705
   9:     0x7ff6bc9e4e1d - std::panicking::begin_panic_handler::closure$0
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\panicking.rs:597
  10:     0x7ff6bc9e2a69 - std::sys_common::backtrace::__rust_end_short_backtrace<std::panicking::begin_panic_handler::closure_env$0,never$>
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\sys_common\backtrace.rs:151
  11:     0x7ff6bc9e4b20 - std::panicking::begin_panic_handler
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\panicking.rs:593
  12:     0x7ff6bcc86815 - core::panicking::panic_fmt
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\core\src\panicking.rs:67
  13:     0x7ff6bcc86e23 - core::result::unwrap_failed
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\core\src\result.rs:1651
  14:     0x7ff6bc657e40 - enum2$<core::result::Result<tuple$<>,anyhow::Error> >::unwrap<tuple$<>,anyhow::Error>
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be\library\core\src\result.rs:1076
  15:     0x7ff6ba56bb1b - dioxus_cli::builder::build::closure$0
                               at C:\Users\aurel\source\repos\dioxus_forks\dioxus\packages\cli\src\builder.rs:118
  16:     0x7ff6ba70ea16 - std::panicking::try::do_call<dioxus_cli::builder::build::closure_env$0,tuple$<> >
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be\library\std\src\panicking.rs:500
  17:     0x7ff6ba712d93 - std::panicking::try::do_catch<core::panic::unwind_safe::AssertUnwindSafe<tokio::runtime::task::harness::poll_future::closure_env$0<futures_util::future::future::Map<futures_u
til::future::try_future::MapErr<hyper::client::conn::Connection<enum2$<hyper_rust
  18:     0x7ff6ba70c1b4 - std::panicking::try<tuple$<>,dioxus_cli::builder::build::closure_env$0>
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be\library\std\src\panicking.rs:464
  19:     0x7ff6ba6daba9 - std::panic::catch_unwind<dioxus_cli::builder::build::closure_env$0,tuple$<> >
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be\library\std\src\panic.rs:142
  20:     0x7ff6ba4968ca - dioxus_cli::builder::build
                               at C:\Users\aurel\source\repos\dioxus_forks\dioxus\packages\cli\src\builder.rs:114
  21:     0x7ff6ba3c0a99 - dioxus_cli::server::web::serve::async_fn$0
                               at C:\Users\aurel\source\repos\dioxus_forks\dioxus\packages\cli\src\server\web\mod.rs:95
  22:     0x7ff6ba3c2445 - dioxus_cli::server::web::startup::async_fn$0
                               at C:\Users\aurel\source\repos\dioxus_forks\dioxus\packages\cli\src\server\web\mod.rs:82
  23:     0x7ff6ba242f03 - dioxus_cli::cli::serve::impl$0::serve::async_fn$0
                               at C:\Users\aurel\source\repos\dioxus_forks\dioxus\packages\cli\src\cli\serve.rs:49
  24:     0x7ff6ba37e49f - dx::main::async_block$0
                               at C:\Users\aurel\source\repos\dioxus_forks\dioxus\packages\cli\src\main.rs:84
  25:     0x7ff6ba23a6f0 - tokio::runtime::park::impl$4::block_on::closure$0<enum2$<dx::main::async_block_env$0> >
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\park.rs:282
  26:     0x7ff6ba23a4d0 - tokio::runtime::coop::with_budget
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\coop.rs:107
  27:     0x7ff6ba23a4d0 - tokio::runtime::coop::budget
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\coop.rs:73
  28:     0x7ff6ba23a4d0 - tokio::runtime::park::CachedParkThread::block_on<enum2$<dx::main::async_block_env$0> >
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\park.rs:282
  29:     0x7ff6ba2eea9c - tokio::runtime::context::blocking::BlockingRegionGuard::block_on<enum2$<dx::main::async_block_env$0> >
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\context\blocking.rs:66
  30:     0x7ff6ba25aa8f - tokio::runtime::scheduler::multi_thread::impl$0::block_on::closure$0<enum2$<dx::main::async_block_env$0> >
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\scheduler\multi_thread\mod.rs:87
  31:     0x7ff6ba33c4dd - tokio::runtime::context::runtime::enter_runtime<tokio::runtime::scheduler::multi_thread::impl$0::block_on::closure_env$0<enum2$<dx::main::async_block_env$0> >,enum2$<core::re
sult::Result<tuple$<>,anyhow::Error> > >
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\context\runtime.rs:65
  32:     0x7ff6ba25aa3a - tokio::runtime::scheduler::multi_thread::MultiThread::block_on<enum2$<dx::main::async_block_env$0> >
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\scheduler\multi_thread\mod.rs:86
  33:     0x7ff6ba2dbd90 - tokio::runtime::runtime::Runtime::block_on<enum2$<dx::main::async_block_env$0> >
                               at C:\Users\aurel\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\runtime\runtime.rs:349
  34:     0x7ff6ba204310 - dx::main
                               at C:\Users\aurel\source\repos\dioxus_forks\dioxus\packages\cli\src\main.rs:69
  35:     0x7ff6ba27efeb - core::ops::function::FnOnce::call_once<enum2$<core::result::Result<tuple$<>,anyhow::Error> > (*)(),tuple$<> >
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be\library\core\src\ops\function.rs:250
  36:     0x7ff6ba2f29ee - std::sys_common::backtrace::__rust_begin_short_backtrace<enum2$<core::result::Result<tuple$<>,anyhow::Error> > (*)(),enum2$<core::result::Result<tuple$<>,anyhow::Error> > >  
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be\library\std\src\sys_common\backtrace.rs:135
  37:     0x7ff6ba243e51 - std::rt::lang_start::closure$0<enum2$<core::result::Result<tuple$<>,anyhow::Error> > >
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be\library\std\src\rt.rs:166
  38:     0x7ff6bc9cfea8 - std::rt::lang_start_internal::closure$2
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library\std\src\rt.rs:148
  44:     0x7ff6ba2043f9 - main
  45:     0x7ff6bcc82b9c - invoke_main
                               at D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  46:     0x7ff6bcc82b9c - __scrt_common_main_seh
                               at D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
  47:     0x7ffa78967344 - BaseThreadInitThunk
  48:     0x7ffa793026b1 - RtlUserThreadStart
Error: 🚫 Serving project failed: Build Failed: Bindgen build failed!
This is probably due to the Bindgen version, dioxus-cli using `0.2.81` Bindgen crate.
error: process didn't exit successfully: `target\debug\dx.exe serve` (exit code: 1)
```

Reply:
I forgot to mention the working directories for each command:
```
cargo run --package dioxus-playwright-liveview-test --bin dioxus-playwright-liveview-test
# playwright-tests/web
cargo run --package dioxus-cli -- serve
# playwright-tests/fullstack
cargo run --package dioxus-cli -- build --features web --release\ncargo run --release --features ssr
```

With those modifications each of the commands works on my windows device, but playwright itself fails

Reply:
https://github.com/DioxusLabs/dioxus/pull/1480 should fix this